src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1101,6 +1101,22 @@ private ImmutableArray<Diagnostic> FilterDiagnosticsSuppressedInSourceOrByAnalyz
             return builder.ToImmutable();
         }
 
+        internal ImmutableArray<Diagnostic> ApplyProgrammaticSuppressionsAndFilterDiagnostics(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            if (reportedDiagnostics.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            var diagnostics = ApplyProgrammaticSuppressions(reportedDiagnostics, compilation);
+            if (compilation.Options.ReportSuppressedDiagnostics || diagnostics.All(d => !d.IsSuppressed))
```

Comment:
>|| diagnostics.All(d => !d.IsSuppressed) [](http://example.com/codeflow?start=64&length=40)

Is this check necessary? (`WhereAsArray()` below will return the original array without allocations if all items satisfy the predicate.)